providers/azure.go

```diff
@@ -149,13 +145,103 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 	created := time.Now()
 	expires := time.Unix(jsonResponse.ExpiresOn, 0)
 
-	return &sessions.SessionState{
+	session := &sessions.SessionState{
 		AccessToken:  jsonResponse.AccessToken,
 		IDToken:      jsonResponse.IDToken,
 		CreatedAt:    &created,
 		ExpiresOn:    &expires,
 		RefreshToken: jsonResponse.RefreshToken,
-	}, nil
+	}
+
+	email, err := p.verifyTokensAndExtractEmail(ctx, session.IDToken, session.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	session.Email = email
+
+	return session, nil
+}
+
+// EnrichSession finds the email to enrich the session state
+func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
+	}
+
+	email, err := getEmailFromProfileAPI(ctx, s.AccessToken, p.ProfileURL.String())
+	if email == "" || err != nil {
+		if err != nil {
+			err = fmt.Errorf("unable to get email address, error: %v", err)
+		} else {
+			err = errors.New("unable to get email address")
+		}
+		logger.Errorf("%s", err)
+		return err
+	}
+	s.Email = email
+
+	return nil
+}
+
+func (p *AzureProvider) prepareRedeem(redirectURL, code string) (url.Values, error) {
+	params := url.Values{}
+	if code == "" {
+		return params, ErrMissingCode
+	}
+	clientSecret, err := p.GetClientSecret()
+	if err != nil {
+		return params, err
+	}
+
+	params.Add("redirect_uri", redirectURL)
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", clientSecret)
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		params.Add("resource", p.ProtectedResource.String())
+	}
+	return params, nil
+}
+
+// verifyTokensAndExtractEmail tries to extract email claim from id_token first
+// and falls back to access token due to
+// https://github.com/oauth2-proxy/oauth2-proxy/pull/914#issuecomment-782285814
+// https://github.com/AzureAD/azure-activedirectory-library-for-java/issues/117
+func (p *AzureProvider) verifyTokensAndExtractEmail(ctx context.Context, idToken, accessToken string) (string, error) {
+	email := ""
+	if p.Verifier == nil {
+		return "", nil
+	}
+
+	if idToken != "" {
+		token, err := p.Verifier.Verify(ctx, idToken)
```

Comment:
These two snippets look almost identical - can we just make a generic verify + extract claims from token methods and call it twice (once for id_token, fallback to access token)

Reply:
do you think this is necessary? If I refactor this way, I will have to break down the calling function to appease the static analysis :-/

Reply:
We can leave it if you are having trouble with the codeclimate checks.

The current construction is just a little hard to read - it has way too many nested if/else loops than we prefer in the project's general code style. We aim for using "early returns" & helper methods to reduce deep nesting of conditional logic.

Unrelated: these `logger.Printf` in the error cases should be `logger.Errorf`, right?

Reply:
nvm, found a way to refactor this!